nomos-libp2p/src/behaviour/nat/address_mapper/nat_pmp.rs

```diff
@@ -0,0 +1,158 @@
+    where
+        Self: Sized,
+    {
+        let nat_pmp = new_tokio_natpmp()
+            .await
+            .map_err(|e| AddressMapperError::PortMappingFailed(e.to_string()))?;
+
+        Ok(Box::new(Self { settings, nat_pmp }))
+    }
+
+    async fn map_address(&mut self, address: &Multiaddr) -> Result<Multiaddr, AddressMapperError> {
+        let (port, protocol) = extract_port_and_protocol(address)?;
+
+        self.send_map_request(protocol, port).await?;
+
+        let public_port = timeout(self.settings.timeout, self.recv_map_response_public_port())
+            .await
+            .map_err(|_| {
+                AddressMapperError::PortMappingFailed(
+                    "Timeout waiting for NAT-PMP mapping response".to_owned(),
+                )
+            })??;
+
+        let public_ip = timeout(self.settings.timeout, self.query_public_ip())
+            .await
+            .map_err(|_| {
+                AddressMapperError::PortMappingFailed(
+                    "Timeout waiting for NAT-PMP public IP response".to_owned(),
+                )
+            })??;
+
+        public_address(address, public_ip, public_port)
+    }
+}
+
+fn extract_port_and_protocol(addr: &Multiaddr) -> Result<(u16, Protocol), AddressMapperError> {
+    addr.iter()
+        .find_map(|p| match p {
+            MaProto::Tcp(p) => Some((p, Protocol::TCP)),
+            MaProto::Udp(p) => Some((p, Protocol::UDP)),
+            _ => None,
+        })
+        .ok_or_else(|| {
+            AddressMapperError::MultiaddrParseError(
+                "No TCP or UDP port found in multiaddr".to_owned(),
+            )
+        })
+}
+
+pub fn public_address(
+    internal: &Multiaddr,
+    public_ip: Ipv4Addr,
+    public_port: u16,
+) -> Result<Multiaddr, AddressMapperError> {
+    let with_ip = internal
+        .replace(0, |_| Some(multiaddr::Protocol::Ip4(public_ip)))
+        .ok_or_else(|| {
+            AddressMapperError::MultiaddrParseError("No IP address found in multiaddr".to_owned())
+        })?;
+
+    with_ip
+        .replace(1, |p| match p {
+            multiaddr::Protocol::Tcp(_) => Some(multiaddr::Protocol::Tcp(public_port)),
+            multiaddr::Protocol::Udp(_) => Some(multiaddr::Protocol::Udp(public_port)),
+            _ => None,
+        })
+        .ok_or_else(|| {
+            AddressMapperError::MultiaddrParseError(
+                "No TCP or UDP port found in multiaddr".to_owned(),
+            )
+        })
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    #[ignore = "Needs real NAT-PMP gateway to test"]
+    async fn test_nat_pmp() {
+        let mut natpmp = NatPmp::initialize(NatMappingSettings {
+            lease_duration: 3600,
+            timeout: std::time::Duration::from_secs(5),
+        })
+        .await
+        .unwrap();
+
+        let addr: Multiaddr = "/ip4/127.0.0.1/tcp/54321".parse().unwrap();
```

Comment:
Same as my other comment. This test has very little meaning if we leave it like this. Either we mock it, if possible, or we simply leave it to be run and capture its output.